src/ec2/model.ts

```diff
@@ -41,25 +41,33 @@ export class Ec2ConnectionManager {
     }
 
     protected async getAttachedPolicies(instanceId: string): Promise<IAM.attachedPoliciesListType> {
-        try {
-            const IamRole = await this.ec2Client.getAttachedIamRole(instanceId)
-            const iamResponse = await this.iamClient.listAttachedRolePolicies(IamRole!.Arn!)
-            return iamResponse.AttachedPolicies!
-        } catch (err) {
+        const IamRole = await this.ec2Client.getAttachedIamRole(instanceId)
+        if (!IamRole) {
             return []
         }
+        const iamResponse = await this.iamClient.listAttachedRolePolicies(IamRole!.Arn!)
+
+        return iamResponse.AttachedPolicies ?? []
     }
 
     public async hasProperPolicies(instanceId: string): Promise<boolean> {
         const attachedPolicies = (await this.getAttachedPolicies(instanceId)).map(policy => policy.PolicyName!)
         const requiredPolicies = ['AmazonSSMManagedInstanceCore', 'AmazonSSMManagedEC2InstanceDefaultPolicy']
 
-        return requiredPolicies.every(policy => attachedPolicies.includes(policy))
+        return requiredPolicies.length !== 0 && requiredPolicies.every(policy => attachedPolicies.includes(policy))
     }
 
-    public async handleStartSessionError(err: AWSError, selection: Ec2Selection): Promise<Error> {
-        const isInstanceRunning = (await this.ec2Client.getInstanceStatus(selection.instanceId)) == 'running'
+    private async isInstanceConnectable(instanceId: string): Promise<boolean> {
+        const isInstanceRunning = (await this.ec2Client.getInstanceStatus(instanceId)) == 'running'
+        const hasProperPolicies = await this.hasProperPolicies(instanceId)
+
+        return isInstanceRunning && hasProperPolicies
+    }
```

Comment:
It would be helpful to return (or throw) _why_ the instance cannot be connected to.

Reply:
Right now I have these as two separate methods, but it sounds like it makes more sense to combine them into one.